### PR TITLE
net: improve max-connection limits code

### PR DESCRIPTION
**Pull request by mzumsande**
This is joint work with amitiuttarwar.

This has the first few commits of #28463. It is not strictly a prerequisite for that, but has changes that in our opinion make sense on their own.
It improves the handling of maximum numbers for different connection types (that are set during init and don’t change after) by:
* moving all calculations into one place, `CConnMan::Init()`. Before, they were dispersed between `Init`, `CConnman::Init` and other parts of `CConnman`, resulting in some duplicated test code.
* removing the possibility of having a negative maximum of inbound connections, which is hard to argue about
* renaming of variables and doc improvements

### CODE REVIEW

File: src/net.h [hunk 1]
```diff
@@ -1197,8 +1197,19 @@ class CConnman
 
     std::unique_ptr<CSemaphore> semOutbound;
     std::unique_ptr<CSemaphore> semAddnode;
+
+    /**
+     * Maximum number of automatic connections permitted, excluding manual
+     * connections but including inbounds. May be changed by the user and is
+     * potentially limited by the number of available file descriptors.
```
Comment (naumenkogs) [inline src/net.h:1204]: 96904e8df098ba35cc5b9b8a089e472f729c0b6a

nit: Perhaps I'd be more generic saying something-something operating system limits

**Comment by naumenkogs** [comment]
ACK 96904e8df098ba35cc5b9b8a089e472f729c0b6a

nice catching the negative number. I haven't found a way to exploit this bug.

File: src/net.h (hunk 1 above)
Comment (mzumsande) [reply src/net.h]: I've mentioned both reasons now (the generic and the specific one).

**Comment by mzumsande** [comment]
[96904e8 ](https://github.com/bitcoin/bitcoin/commit/96904e8df098ba35cc5b9b8a089e472f729c0b6a)to df69b22: Rebased, and changed a comment slightly.

File: src/init.cpp [hunk 2]
```diff
@@ -489,7 +489,7 @@ void SetupServerArgs(ArgsManager& argsman)
...
     argsman.AddArg("-listen", strprintf("Accept connections from outside (default: %u if no -proxy, -connect or -maxconnections=0)", DEFAULT_LISTEN), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-listenonion", strprintf("Automatically create Tor onion service (default: %d)", DEFAULT_LISTEN_ONION), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    argsman.AddArg("-maxconnections=<n>", strprintf("Maintain at most <n> connections to peers (default: %u). This limit does not apply to connections manually added via -addnode or the addnode RPC, which have a separate limit of %u.", DEFAULT_MAX_PEER_CONNECTIONS, MAX_ADDNODE_CONNECTIONS), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-maxconnections=<n>", strprintf("Maintain at most <n> automatic connections to peers (default: %u). This limit does not apply to connections manually added via -addnode or the addnode RPC, which have a separate limit of %u.", DEFAULT_MAX_PEER_CONNECTIONS, MAX_ADDNODE_CONNECTIONS), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
Comment (naumenkogs) [inline src/init.cpp:492]: df69b22f2e3cc03764a582f29a16a36114f67e17

have you considered changing the named argument? `-maxconnections` -> `-maxautoconnections`.
It should in that case be backward-compatible (still allow legacy `-maxconnections`)

**Comment by naumenkogs** [comment]
ACK df69b22f2e3cc03764a582f29a16a36114f67e17

**Comment by achow101** [comment]
ACK df69b22f2e3cc03764a582f29a16a36114f67e17

File: src/init.cpp (hunk 2 above)
Comment (mzumsande) [reply src/init.cpp:492]: I forgot about this comment - I hadn't considered it, and I think I would prefer `maxautoconnections`, but changing it (even with keeping an alias) seems like too much effort for too little benefit for my taste .

**Pull request merged**